### ISSUE TITLE
IPv6 Options: "Allow IPv6" is not persistent after reboot. 

### ISSUE THREAD

**Issue by giovino** [opnsense/core#490]
Version:
```
OPNsense 15.7.20-amd64	
FreeBSD 10.1-RELEASE-p24	
OpenSSL 1.0.2d 9 Jul 2015
```

It does not appear that the configuration ```IPv6 Options: "Allow IPv6"``` is persistent after reboot. I am able to replicate this by rebooting the machine and then witnessing IPv6 traffic being blocked in the firewall logs. If I then un-check "allow ipv6" -> save - re-check "allow ipv6" -> save... Outbound IPv6 traffic is once again allowed. 

See the following [forum thread](https://forum.opnsense.org/index.php?topic=1777.0) for more information.

**Comment by fichtner** [comment]
This might be a side-effect of something else, I don't think it's generally not working as that would mean no IPv6 for everybody. :) As a first step, I'm going to need /tmp/rules.debug directly after reboot (not working) and then after reload (working). You can send them privately via the forum or email. Thanks so far. If the two are equal, we are dealing with a deeper issue.

As a side note, IPv4 works fine regardless?

**Comment by giovino** [comment]
Files sent to you via email to your opnsense.org email account. 

They are not equal.. hopefully that is good :)

I've not experienced any issues with IPv4 to my knowledge.

**Comment by fichtner** [comment]
Okay, looks consistent. What happens is that your LAN IPv6 is not assigned during bootup, that's why the initial ruleset fails. It will block access from LAN to your DHCPv6, rendering your IPv6 clients unable to connect. Is that correct: are you running a DHCPv6 on the OPNsense LAN which is configured via track WAN?

**Comment by giovino** [comment]
Probably :)

```
Interfaces: WAN
IPv4 Configuration Type: DHCP
IPv6 Configuration Type: DHCPv6
DHCPv6 Prefix Delegation size: 64
Send IPv6 prefix hint: <checked>
```

```
Interfaces: LAN
IPv4 Configuration Type: Static IPv4
IPv6 Configuration Type: Track Interface
Static IPv4 Configuration
 - IPv4 address: 192.168.1.1/24
Track IPv6 Interface
 - IPv6 Interface: WAN
```

I use [this article](https://www.kirkg.us/posts/setting-up-ipv6-with-opnsense-and-comcast/) as a reference. 

Should I be configuring this differently?

**Comment by fichtner** [comment]
No, that's correct. It's just that your LAN (or perhaps also WAN) are not available directly on boot, maybe just take a little too long to get assigned. I will report back tomorrow.

**Comment by giovino** [comment]
I could imagine that is the case. I am using a [Jetway JBC200F9N-E4IN-B](http://www.jetwaycomputer.com/JBC200F9N-E4IN.html) with a [Jetway ADE4INLANG GEN2](http://www.jetwaycomputer.com/Expansion.html) Daughter Board. I can imagine some nic's not being available directly on boot. 

I've sent you an email with a file with dmesg.today output that may help.

**Comment by fichtner** [comment]
It's not about the NIC, more about the WAN connection / upstream DHCPv6. What happens when you unplug your WAN and plug it back in? Is IPv6 blocked again (like clients get no more IPv6)?

As far as code is concerned, some initialisations are skipped when booting, but this still seems kind of unique to your setup / timing.  We can probably add a workaround, but only when sure there is no other way to get your setup to run as it should out of the box. :)

**Comment by giovino** [comment]
> What happens when you unplug your WAN and plug it back in? Is IPv6 blocked again (like clients get no more IPv6)?

I unplugged the WAN ethernet cable for 10+ seconds and then plugged it back in. Outbound IPv6 LAN traffic was **not** blocked by the firewall.  (i.e. it worked correctly)

**Comment by fichtner** [comment]
Ok, thanks. Well, the outbound traffic is not really the problem. What is being blocked is the IPv6 address to listen to for DHCPv6 leases on the LAN side. If your client still had an IP it's unlikely it would have connectivity issues when starting up. Can you make sure that after plugging it back in you can get a lease with a LAN client?

**Comment by giovino** [comment]
That explanation makes sense, I was confused by the diff in the firewall rule set. 

On an OS X host, I did a few things to get OS X to request new DHCPv6 lease:

1. Turn off Wifi -> Turn on Wifi
2. Set ```Configure IPv6```: ```Link-local only``` ->  ```Automatically```

In both cases IPv6 connectivity worked as expected and/or it appears that the host on the LAN was able to listen for DHCPv6 leases. (i.e. it worked correctly)

**Comment by fichtner** [comment]
And Internet connectivity works in this case or not? Link-local might not be enough.

**Comment by giovino** [comment]
```Configure IPv6: Link-local only``` results in:

```
$ ping6 google.com
ping6: UDP connect: No route to host
```

Setting ``Configure IPv6``` back to ```Automatically```:

```
$ ping6 google.com
PING6(56=40+8+8 bytes) <ipv6 address> --> 2607:f8b0:4009:808::200e
16 bytes from 2607:f8b0:4009:808::200e, icmp_seq=0 hlim=54 time=24.132 ms
16 bytes from 2607:f8b0:4009:808::200e, icmp_seq=1 hlim=54 time=23.304 ms
```

**Comment by fichtner** [comment]
Ah, ok, gotcha... so it points to only bootup is preventing IPv6 connectivity. Thanks so far. One last look at the code...

**Comment by fichtner** [comment]
@giovino and me agreed to close this after more testing. It was fixed in 15.7.22.
